marker_api/src/sem/ty.rs

```diff
@@ -154,3 +203,183 @@ macro_rules! impl_ty_data {
     };
 }
 use impl_ty_data;
+
+/// This struct describes a trait and its generic arguments. It is used to check
+/// if a semantic type implements a trait.
+///
+/// See [`TyKind::implements_trait`].
+#[derive(Debug)]
+pub struct TestTraitRef {
+    trait_ref: TyDefIdSource,
+    // TODO generics
+    mode: TraitTestMode,
+}
+
+impl TestTraitRef {
+    pub fn builder() -> TestTraitRefBuilder {
+        TestTraitRefBuilder::new()
+    }
+}
+
+/// A builder used to construct a [`TestTraitRef`] instance.
+#[derive(Debug)]
+pub struct TestTraitRefBuilder {
+    trait_ref: Option<TyDefIdSource>,
+    mode: TraitTestMode,
+}
+
+impl TestTraitRefBuilder {
+    fn new() -> Self {
+        Self {
+            trait_ref: None,
+            mode: TraitTestMode::default(),
+        }
+    }
+
+    pub fn trait_from_path(&mut self, path: impl Into<String>) -> &mut Self {
+        self.trait_ref = Some(TyDefIdSource::Path(path.into()));
+        self
+    }
+
+    pub fn trait_from_id(&mut self, id: impl Into<TyDefId>) -> &mut Self {
+        self.trait_ref = Some(TyDefIdSource::Id(id.into()));
+        self
+    }
+
+    pub fn mode(&mut self, mode: TraitTestMode) -> &mut Self {
+        self.mode = mode;
+        self
+    }
+
+    pub fn build(&mut self) -> TestTraitRef {
+        TestTraitRef {
+            trait_ref: self.trait_ref.take().expect("the trait was never set"),
+            mode: self.mode,
+        }
+    }
+}
+
+#[derive(Debug)]
+#[cfg_attr(feature = "driver-api", visibility::make(pub))]
+pub(crate) struct FfiTestTraitRef<'a> {
+    trait_ids: FfiSlice<'a, TyDefId>,
+    // TODO generics
+    mode: TraitTestMode,
+}
+
+#[cfg(feature = "driver-api")]
+impl<'a> FfiTestTraitRef<'a> {
+    pub fn trait_ids(&self) -> &'a [TyDefId] {
+        self.trait_ids.get()
+    }
+
+    pub fn mode(&self) -> TraitTestMode {
+        self.mode
+    }
+}
+
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum TyDefIdSource {
+    Id(TyDefId),
+    // TODO: Handle the path properly, since `Str` is not ABI safe
+    Path(String),
+}
+
+impl From<TyDefId> for TyDefIdSource {
+    fn from(value: TyDefId) -> Self {
+        TyDefIdSource::Id(value)
+    }
+}
+
+impl From<String> for TyDefIdSource {
+    fn from(value: String) -> Self {
+        TyDefIdSource::Path(value)
+    }
+}
+
+/// This enum defines how strict the [`TyKind::implements_trait`] test should be.
+/// The difference is probably best illustrated by examples. For that, let's
+/// take the following traits:
+///
+/// ```
+/// // Definitions
+/// trait SimpleTrait { /* ... */ }
+/// trait GenericTrait<T> { /* ... */ }
+/// trait TwoGenericTrait<T, U=u8> { /* ... */ }
```

Comment:
There is also a case for const generics in traits, although I that's a rare case and we can implement support for that later